File: backend/src/main/java/de/mrsebastian/todoappdemo/backend/task/service/TaskMapper.java
package de.mrsebastian.todoappdemo.backend.task.service;

import de.mrsebastian.todoappdemo.backend.task.dataaccess.TaskCreateDao;
import de.mrsebastian.todoappdemo.backend.task.dataaccess.TaskDao;
import de.mrsebastian.todoappdemo.backend.task.dataaccess.TaskUpdateDao;
import de.mrsebastian.todoappdemo.task.api.rest.model.TaskCreateDTO;
import de.mrsebastian.todoappdemo.task.api.rest.model.TaskDTO;
import de.mrsebastian.todoappdemo.task.api.rest.model.TaskUpdateDTO;
import org.mapstruct.Mapper;

@Mapper
public interface TaskMapper {

    TaskCreateDao toEntity(TaskCreateDTO dto);

    TaskUpdateDao toUpdateDao(TaskUpdateDTO dto);

    TaskDTO toDTO(TaskDao taskDao);
}

File: frontend/frontend/src/features/task/services/api/model/TaskMapper.ts
import TaskCreateDTO from "@/features/task/services/api/model/TaskCreateDTO";
import TaskDTO from "@/features/task/services/api/model/TaskDTO";
import TaskUpdateDTO from "@/features/task/services/api/model/TaskUpdateDTO";
import Task from "@/features/task/types/Task";
import TaskPersisted from "@/features/task/types/TaskPersisted";

export default class TaskMapper {
    taskToWriteDto(task: Task): TaskCreateDTO {
        return new TaskCreateDTO(task.title, task.description, task.dueDate);
    }

    taskDtoToTaskPersisted(dto: TaskDTO): TaskPersisted {
        return new TaskPersisted(
            dto.id,
            dto.title,
            dto.description,
            dto.dueDate
        );
    }

    persistedTaskToTaskUpdateDTO(persistedTask: TaskPersisted): TaskUpdateDTO {
        return new TaskUpdateDTO(
            persistedTask.title,
            persistedTask.description ?? "",
            persistedTask.dueDate
        );
    }
}
